### PR TITLE
Remove faiss dependency from fused_l2_knn.cuh, selection_faiss.cuh, ball_cover.cuh and haversine_distance.cuh

### PR DESCRIPTION
**Pull request by benfred**
Remove the dependency on faiss from the fused_l2_knn.cuh, selection_faiss.cuh, ball_cover.cuh and haversine_distance.cuh headers.

This takes a copy of the faiss BlockSelect/WarpSelect device code for top-k selection, and updates to use
raft primitives for things like reductions,  KeyValuePair, warp shuffling etc.

### CODE REVIEW

**Review by cjnolet — APPROVED** [review]
Changes look great!

**Review by raydouglass — APPROVED** [review]

File: cpp/include/raft/spatial/knn/detail/faiss_select/Comparators.cuh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
```
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/include/raft/spatial/knn/detail/faiss_select/Comparators.cuh:5]: Thanks for pointing this out, @bdice. Yes, a file `LICENSE.faiss` needs to be added to `raft/thirdparty/LICENSES`. We did this for [cuML](https://github.com/rapidsai/cuml/blob/branch-23.02/thirdparty/LICENSES/LICENSE.faiss) and I didn't realize it hadn't yet been done for RAFT.

[the remark this answers is not in the document]
Comment (benfred) [reply cpp/include/raft/spatial/knn/detail/faiss_select/Comparators.cuh]: Thanks @bdice! I've updated these headers like @cjnolet suggested - to match how its happening with cuml

**Comment by benfred** [comment]
@gpucibot merge

**Pull request merged**